src/Wind.cpp

```diff
@@ -0,0 +1,120 @@
+    parameters.length_overall = utilities::getParameter<double>("Wind", el, "length_overall", "m", 0);
+    parameters.air_density = utilities::getParameter<double>("Wind", el, "air_density", "kg/m3", 1.12);
+    parameters.coefficients = utilities::getParameter<Vector3d>("Wind", el, "wind_coeffs",
+                                                                "", Vector3d(0, 0, 0));
+    return parameters;
+}
+
+void Wind::readWindVelocity(const ConstVector3dPtr &velocity)
+{
+    mWindVelocity_ned = Vector3d(velocity->x(), velocity->y(), velocity->z());
+}
+
+Wind::Effects Wind::computeEffects(Quaterniond const body2world_orientation_ned, Vector3d const vessel_linear_vel_world_ned, Vector3d wind_velocity_world_ned)
+{
+    Quaterniond world2body_orientation = body2world_orientation_ned.Inverse();
+    Vector3d relative_wind_velocity_world = vessel_linear_vel_world_ned - wind_velocity_world_ned;
+    Vector3d relative_wind_velocity_body = world2body_orientation * relative_wind_velocity_world;
+    relative_wind_velocity_body.Z() = 0;
+    if (!relative_wind_velocity_body.X() && !relative_wind_velocity_body.Y())
+        return Effects{};
+
+    // Compute the wind's angle of attack and its coefficients
+    double angle_of_attack_rad = -atan2(relative_wind_velocity_body.Y(), relative_wind_velocity_body.X());
+    Angle angle_of_attack(angle_of_attack_rad);
+    double wind_coeff_x = -mParameters.coefficients.X() * cos(angle_of_attack.Radian());
+    double wind_coeff_y = mParameters.coefficients.Y() * sin(angle_of_attack.Radian());
+    double wind_coeff_n = mParameters.coefficients.Z() * sin(2 * angle_of_attack.Radian());
+
+    // Compute wind effects for X, Y and N
+    Effects wind_effects;
+    wind_effects.force[0] = 0.5 * mParameters.air_density * relative_wind_velocity_body.SquaredLength() * wind_coeff_x * mParameters.frontal_area;
+    wind_effects.force[1] = 0.5 * mParameters.air_density * relative_wind_velocity_body.SquaredLength() * wind_coeff_y * mParameters.lateral_area;
+    wind_effects.torque[2] = 0.5 * mParameters.air_density * relative_wind_velocity_body.SquaredLength() * wind_coeff_n * mParameters.lateral_area * mParameters.length_overall;
+    return wind_effects;
+}
+
+void Wind::update()
+{
+    // Convert model orientation and velocity to NED coordinate system
+    Quaterniond model_orientation_ned = utilities::NWU2NEDCoordinateSystem<Quaterniond>(mModel->WorldPose().Rot());
+    Vector3d model_velocity_ned = utilities::NWU2NEDCoordinateSystem<Vector3d>(mModel->WorldLinearVel());
+
+    // Compute the new force and torque for this timestep and convert the effects to NWU coordinate system (used by gazebo)
+    Effects effects = computeEffects(model_orientation_ned, model_velocity_ned, mWindVelocity_ned);
+    effects.force = utilities::NED2NWUCoordinateSystem(effects.force);
+    effects.torque = utilities::NED2NWUCoordinateSystem(effects.torque);
```

Comment:
Oh my ...

Nope. Just change the Fossen formulas to be in NWU. All the vector math is unchanged, you only have to change the force/torque.

Which happens to be a VERY simple transformation:
- x = x
- y = -y
- z = -z

Reply:
Ah, these are just remnants from experiments I made to identify the issues with the plugin. I found implementing the conversion like this easier because I wasn't sure at the time which coordinate system the simulation used, so I could change this the ENU by switching the function I called.  

Actually, it isn't necessary to transform the effects to NWU, since the model's pose and velocity are already in the same coordinate system.